src/coreclr/jit/codegencommon.cpp

```diff
@@ -4794,14 +4794,41 @@ void CodeGen::genCheckUseBlockInit()
                 }
             }
 
-            /* With compInitMem, all untracked vars will have to be init'ed */
-            /* VSW 102460 - Do not force initialization of compiler generated temps,
-                unless they are untracked GC type or structs that contain GC pointers */
-            CLANG_FORMAT_COMMENT_ANCHOR;
+            bool mustInitThisVar = false;
+
+            const bool isTemp      = varDsc->lvIsTemp;
+            const bool hasGCPtr    = varDsc->HasGCPtr();
+            const bool isTracked   = varDsc->lvTracked;
+            const bool isStruct    = varTypeIsStruct(varDsc);
+            const bool compInitMem = compiler->info.compInitMem;
 
-            if ((!varDsc->lvTracked || (varDsc->lvType == TYP_STRUCT)) && varDsc->lvOnFrame)
```

Comment:
The new logic doesn't take into account `varDsc->lvOnFrame`. Is that is intentional?

Reply:
Good question, if I add an assert like `if (mustInitThisVar) assert(varDsc->lvOnFrame);` it won't fail (x64 spmi) but it is a current implicit contract so we probably should not rely on it.
```
if (hasGCPtr && !isTracked) -> not tracked means not in register -> lvOnFrame == true
else if (hasGCPtr && isStruct) -> do not enregister structs, with https://github.com/dotnet/runtime/pull/51850/checks?check_run_id=2447290001 won't enregister structs with GC pointers -> lvOnFrame == true
if (!isTracked)  -> not tracked means not in register -> lvOnFrame == true
```

I will return the condition, thank you.